### PR TITLE
NPE when deserializing malformed collections from client

### PR DESCRIPTION
**Pull request by jonmeredith**
See CASSANDRA-18505

### CODE REVIEW

File: src/java/org/apache/cassandra/serializers/ListSerializer.java [hunk 1]
```diff
@@ -104,6 +106,8 @@ public <V> List<T> deserializeForNativeProtocol(V input, ValueAccessor<V> access
             {
                 // We can have nulls in lists that are used for IN values
                 V databb = readValue(input, accessor, offset, version);
+                if (databb == null)
+                    throw new MarshalException("Not enough bytes to read value in list");
                 offset += sizeOfValue(databb, accessor, version);
                 if (databb != null)
```
Comment (maedhroz) [inline src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: nit: This check has become unnecessary?

Comment (maedhroz) [reply src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: The comment above seems to think we should be able to have nulls?

Comment (smiklosovic) [reply src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: Why not to throw that exception from that method? We do not need to throw it explicitly everywhere then.

Comment (maedhroz) [reply src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: I think we need to leave the possibility of a null open for lists on a -1 size.

Comment (maedhroz) [reply src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: For sets and maps, I would almost have a separate `readNonNullValue()` that does throw internally.

Comment (jonmeredith) [reply src/java/org/apache/cassandra/serializers/ListSerializer.java:113]: Went with your suggestion - null in the input seems more likely than corruption so a better, cleaner exception.

**Review by maedhroz — APPROVED** [review]
We could always make a small change to `SerDeserTest` to codify our expectations, catch thew new `MarshalException` etc., but I'll leave that up to you.

**Pull request closed without merging**